Introduce a new `base` module

bal-e opened the pull request:
This introduces a `new_base` module as a potential candidate for replacing `base`.  While it has a similar structure, it has some distinguishing features:

1. It uses concrete types (usually slices and references) in place of `Octs` generic parameters.  The hard-coded types are ergonomic and efficient and simplify the API.

2. It introduces a new parsing architecture which aggressively avoids copying.  The architecture is inspired by the `zerocopy` crate, but allows for partial copying when zero-copy is not possible.  `derive` macros are provided (and heavily used internally) to avoid boilerplate when implementing the parsing traits.

3. It overhauls the message building process so that truncation errors are handled more gracefully and message components (i.e. questions and records) can be built incrementally.  This lays the groundwork for a more efficient client/server architecture where messages are built in-place, even across middleware layers.

4. It introduces `RevName`, a more efficient type for domain names (where labels are stored in reverse order).  Operations on `RevName` tend to be more efficient (e.g. canonical-order comparisons are significantly better).

Here's the current roadmap:

- [x] Define basic DNS types (messages, questions, records, etc.).
- [x] Define low-level parsing and building traits.
- [x] Define `derive` macros for the parsing/building traits.
- [x] Define some record data types (probably those from RFC1035).
- [x] Define the OPT record data, and some EDNS option types.
- [x] Finalize the design of the wire-format parsing traits.
  - [x] Add `derive` macros for these traits.
- [x] Overhaul `net::server` (in a separate PR) on top of `new-base`.

Note: this is a new version of #469 with a less problematic branch name.

bal-e committed:
Add module 'new_base'

bal-e committed:
[new_base] Add module 'name'

bal-e committed:
[new_base/name] Define labels

bal-e committed:
[new_base] Add module 'message'

bal-e committed:
[new_base/name/label] Use 'zerocopy'

bal-e committed:
[new_base] Add module 'parse'

bal-e committed:
[new_base/name] Add module 'parsed'

bal-e committed:
[new_base] Add module 'question'

bal-e committed:
Add module 'record'

bal-e committed:
[new_base] Add high-level parsing traits

bal-e committed:
[new_base/name] Add module 'reversed'

bal-e committed:
[new_base/name/reversed] Implement complex parsing

bal-e committed:
[new_base/name] Add some 'Debug' impls

bal-e committed:
[new_base] Implement parsing for 'Question' and 'Record'

bal-e committed:
[new_base/parse] Update with new question/record types

bal-e committed:
[new_base/name] Fix bugs (thanks clippy)

bal-e committed:
[new_base] Make 'record' a private module

bal-e committed:
Use 'zerocopy' 0.8.5 or newer

It implements 'Hash' for the provided integer types.

bal-e committed:
Add module 'new_rdata' with most RFC 1035 types

bal-e committed:
[new_base] Define 'CharStr'

bal-e committed:
[new_rdata] Define 'Hinfo'

bal-e committed:
[new_rdata/rfc1035] Implement (basic) 'Txt' records

bal-e committed:
[new_rdata/rfc1035] Add 'ParseFrom' impls where missing

bal-e committed:
[new_rdata/rfc1035] Don't use 'zerocopy' around names

bal-e committed:
[new_base/charstr] Impl 'Eq' and 'Debug'

bal-e committed:
[new_base] Add module 'serial'

bal-e committed:
[new_base] Add module 'build'

bal-e committed:
[new_base/build] Add a 'Builder' for DNS messages

bal-e committed:
[new_base/name/reversed] Impl building traits

bal-e committed:
[new_base/build] Add convenience impls for '[u8]'

bal-e committed:
[new_base/question] Impl building traits

bal-e committed:
[new_base/record] Support building and overhaul parsing

bal-e committed:
[new_base/charstr] Support building

bal-e committed:
[new_rdata/rfc1035] Impl building traits

bal-e committed:
[build/builder] Improve unclear documentation

bal-e committed:
[new_rdata/rfc1035] gate 'Ipv4Addr' behind 'std'

bal-e committed:
[new_rdata/rfc1035] Gate more things under 'std'

bal-e committed:
[new_base/build/builder] Remove unnecessary explicit lifetime in impl

bal-e committed:
[new_rdata/rfc1035] Support 'Display' outside 'std'

bal-e committed:
[new_rdata] Inline 'rfc1035' and support 'rfc3596'

bal-e committed:
[new_rdata] Rename submodules with more intuitive names

bal-e committed:
[new_base] Set up basic EDNS support

Instead of 'new_base::opt', EDNS is now granted its own top-level
module.  This matches up well with 'crate::tsig'.

bal-e committed:
[new_base/record] Add trait 'ParseRecordData'

bal-e committed:
[new_base/parse] Make 'Split*' imply 'Parse*'

bal-e committed:
[new_base/record] Add a default for 'parse_record_data()'

bal-e committed:
[new_rdata/basic] Use more capitalization in record data type names

bal-e committed:
[new_rdata] Define enum 'RecordData'

bal-e committed:
[new_rdata/basic] Elide lifetime as per Clippy

bal-e committed:
[new_rdata] Elide lifetimes as per clippy

bal-e committed:
Set up a 'domain-macros' crate

'domain-macros' will provide:

- 'derive' macros for DNS-specific (zero-copy) serialization
- 'derive' macros for building and parsing specialized DNS messages
- 'derive' macros for composing clients and servers

The first use case will replace 'zerocopy'.

bal-e committed:
[new_base/parse] Define 'ParseBytesByRef' for deriving

bal-e committed:
[macros] Define 'ImplSkeleton' and prepare a basic derive

bal-e committed:
Expand 'ParseBytesByRef' and largely finish its derive macro

bal-e committed:
[macros] Add module 'repr' for checking for stable layouts

bal-e committed:
[new_base/parse] Implement '*BytesByRef' for '[T; N]'

bal-e committed:
[macros] Add a derive macro for 'SplitBytesByRef'

'ParseBytesByRef' now requires all implementing types to be unaligned.
Otherwise, padding bytes wouldn't be accounted for properly, e.g. in

    // Has alignment of largest field: 8 bytes
    #[repr(C)]
    pub struct Foo {
        a: u8,
        // 7 bytes of padding here
        b: u64,
    }

The 'derive' can't tell how much padding to use, so it would parse a
'[u8; 9]' as a valid instance of 'Foo'.  Every 'repr(C)' would have to
use 'repr(packed)' too.

bal-e committed:
Use '{Parse,Split}BytesByRef' instead of 'zerocopy'

bal-e committed:
Rename '{Split,Parse}From' to '{Split,Parse}Bytes'

bal-e committed:
[macros] Add derives for '{Split,Parse}Bytes'

bal-e committed:
[macros] Factor out 'new_lifetime()'

bal-e committed:
Use parsing trait derives across the 'new_*' codebase

bal-e committed:
Impl and use derives for 'AsBytes' and 'BuildBytes'

bal-e committed:
[macros] Minor fixes as per clippy

bal-e committed:
[new_base/parse] Fix missing doc link

bal-e committed:
[macros] Fix no-fields output for 'ParseBytes'

bal-e committed:
[new_base/serial] Support measuring unix time

bal-e committed:
[new_edns] Implement DNS cookie support

bal-e committed:
[macros] Factor out struct inspection and building

bal-e committed:
[new_edns] Impl RFC 8914 "Extended DNS errors"

tertsdiepraam commented on src/new_base/record.rs:
```diff
@@ -0,0 +1,420 @@
+
+//--- Parsing from bytes
+
+impl<'a, N, D> SplitBytes<'a> for Record<N, D>
```
Hmm this is mostly the same as `SplitFromMessage` can't we implement that one in terms of this?

bal-e replied:
Unfortunately, both traits are necessary.  `SplitFromMessage` gets used when parsing DNS messages from the wire.  `SplitBytes` is also used when parsing from unknown record data (e.g. encoded in a zone file), where there isn't a containing `Message` and name compression is not allowed.  I do want to add derives to make some of this easier.

tertsdiepraam commented on src/new_rdata/mod.rs:
```diff
@@ -0,0 +1,181 @@
+
+/// Data for an unknown DNS record type.
+#[derive(Debug, AsBytes, BuildBytes, ParseBytesByRef)]
+#[repr(C)]
```
Why `C` and not `transparent`?

bal-e replied:
Left over from the `zerocopy` derives, will fix.

tertsdiepraam commented on macros/src/data.rs:
```diff
@@ -0,0 +1,159 @@
+
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{spanned::Spanned, *};
```
You've used some wildcard imports but I think the policy is to not do that for our LSP-less colleages.

bal-e replied:
There are over 30 imports from `syn` IIRC.  I don't want to flood the `use` statement.

partim replied:
Maybe you can prefix the types with `syn::` when used?

bal-e replied:
I've done a mix of explicit imports for the common declarations and `syn::T` for the uncommon ones.  `impls.rs` has taken the hardest hit, it has 3 lines of common imports.

bal-e committed:
[new_edns] Impl parsing/building for 'EdnsOption'

bal-e commented on src/new_base/question.rs:
```diff
@@ -0,0 +1,145 @@
+        message: &'a Message,
+        range: Range<usize>,
+    ) -> Result<Self, ParseError> {
+        let (qname, rest) = N::split_from_message(message, range.start)?;
+        let (&qtype, rest) = <&QType>::split_from_message(message, rest)?;
```
Yeah, `ParseFromMessage` has a fundamental flaw regarding the range handling.  I'm going to rethink those two traits.

bal-e replied:
I've fixed it!  The message now has to be pre-cut using `Message::slice_to()`, and this ensures that the end of the input range is never crossed.

bal-e committed:
[new_base] Add module 'wire' to replace 'zerocopy'

This ended up collecting a lot of small changes as I tried to get
things to compile.

- All the bytes parsing/building traits have been moved to 'wire'.
- The 'wire::ints' module replaces 'U16' and 'U32' from 'zerocopy'.
- '{Parse,Split}BytesByRef' now support parsing from '&mut'.
- Every derive macro is documented under a re-export in 'wire'.

The remaining contents of 'new_base::{build, parse}' might get moved
into a shared 'message' module at some point.  We'll see.

bal-e commented on src/new_base/parse/question.rs:
```diff
@@ -0,0 +1,148 @@
+
+//--- Impls for 'Option<T>'
+
+impl<T: ParseQuestion> ParseQuestion for Option<T> {
```
`None` is returned if no matching `T` was found in the message.

bal-e committed:
[new_base] Correct docs for build traits

bal-e committed:
[macros] Avoid glob imports where possible

bal-e committed:
[macros/lib.rs] Remove the last 'syn' glob import

bal-e committed:
[new_base/parse] Rework '{Parse,Split}FromMessage'

Instead of passing the input selection as a range for parsing, the
whole message is cut (using 'Message::slice_to()') and only the start
is indicated.  This ensures that we never cross the end of the range.
It also implicitly dictates that compressed names are not allowed to
reference future locations in messages.

In addition, both the parsing traits now use offsets into the message
_contents_ rather than the whole message.  They can avoid 'as_bytes()'
everywhere and have better guarantees of success.  It also ensures the
message header can never be selected for parsing.

bal-e committed:
[new_base/name] Delete unused 'parsed.rs'

bal-e committed:
[new_base/wire] Define 'SizePrefixed'

bal-e committed:
[new_rdata/edns] Implement iteration and formatting

bal-e committed:
[new_base/build] Use 'BuildResult' to ensure 'commit()' is called

This has already caught a missing commit (for 'Question').

bal-e committed:
[new_base/build] Define 'MessageBuilder'

bal-e committed:
[new_base/message] Add 'as_bytes_mut()' and 'slice_to_mut()'

bal-e committed:
[new_base/name] Add 'UnparsedName'

bal-e committed:
[new_base/message] Impl 'as_array()' for 'SectionCounts'

bal-e committed:
[new_base/build] Define 'RecordBuilder'

bal-e committed:
[new_base/name] Accept Clippy simplifications

bal-e committed:
Merge branch 'main' into new-base

bal-e committed:
Merge branch 'main' into new-base

bal-e committed:
[new_base/build/record] Track record data size

bal-e committed:
[new_base/parse] Remove 'ParseMessage' etc.

These interfaces need to be redesigned to be more specific to important
use cases.

bal-e committed:
[new_base/build] Document 'BuildCommitted' thoroughly

bal-e committed:
[new_base/build/builder] Rewrite with a lot of documentation

bal-e committed:
[new_base/build] Add module documentation

bal-e committed:
[new_base/parse] Add a bit of module documentation

bal-e committed:
[new_base/parse] Add missing doc links

bal-e committed:
[new_base/wire/ints] Make 'set()' non-const

Philip-NLnetLabs commented on src/new_base/name/label.rs:
```diff
@@ -0,0 +1,279 @@
+    fn split_bytes(bytes: &'a [u8]) -> Result<(Self, &'a [u8]), ParseError> {
+        let (&size, rest) = bytes.split_first().ok_or(ParseError)?;
+        if size < 64 && rest.len() >= size as usize {
+            let (label, rest) = rest.split_at(size as usize);
```
What happens if rest is less than size bytes?

bal-e replied:
In that case, the input was apparently too small to contain the whole label, so a parse error is returned.  You can hit it with e.g. `split_bytes(b"\x07www")`.

Philip-NLnetLabs replied:
Ah, quite right.  I got confused somewhere.

Philip-NLnetLabs commented on src/new_base/question.rs:
```diff
@@ -0,0 +1,197 @@
+    fn parse_from_message(
+        message: &'a Message,
+        start: usize,
+    ) -> Result<Self, ParseError> {
```
Why not call split_from_message and check that rest is empty?

bal-e replied:
While that would be a valid implementation, I am hoping to define new proc macros for `SplitFromMessage` and `ParseFromMessage` soon (in the same fashion as `SplitBytes` and `ParseBytes`).  I chose to define this method with a bit more boilerplate so that I can refer to it as a template when writing the proc macro.

Philip-NLnetLabs commented on src/new_base/question.rs:
```diff
@@ -0,0 +1,197 @@
+#[derive(Clone, BuildBytes, ParseBytes, SplitBytes)]
+pub struct Question<N> {
+    /// The domain name being requested.
+    pub qname: N,
```
I think for names there needs to be some underlying trait that specifies that N can parse and build compressed DNS names. Otherwise, for example, compressed and uncompressed names can get confused or other confusion may arise.

bal-e replied:
It is strongly expected that any domain name type (at least those in this crate) will implement `ParseFromMessage`, `SplitFromMessage` and `BuildIntoMessage` using name (de)compression.  It is currently possible to use an incorrect type here, but I believe the risks of such a confusion are quite low.  In the future, if we introduce new name types which do not follow these conventions, we can introduce e.g. a `Name` trait.

ximon18 replied:
Maybe document these expectations/intentions?

Philip-NLnetLabs commented on src/new_base/wire/parse.rs:
```diff
@@ -0,0 +1,510 @@
+///
+/// [`SplitBytes`] can be derived on `struct`s (not `enum`s or `union`s).  All
+/// fields except the last must implement [`SplitBytes`], while the last field
+/// only needs to implement [`SplitBytes`].
```
Two times SplitBytes?

bal-e replied:
Oops, that was left over when I copied the `ParseBytes` documentation.  Will fix.

Philip-NLnetLabs commented on src/new_edns/cookie.rs:
```diff
@@ -0,0 +1,244 @@
+#[cfg(all(feature = "std", feature = "siphasher"))]
+use crate::new_base::wire::{AsBytes, TruncationError};
+
+//----------- CookieRequest --------------------------------------------------
```
I find this very confusing. There are two types of cookies, short cookies that only contain the client's cookie and longer cookies that contain both the client and the server parts. There are two roles, the client and the server.

It is unclear to me what is what and who does what.

bal-e replied:
A `CookieRequest` is a "client coookie", which requests a `Cookie` ("server cookie") from the server.  I tried these names out as I found them more intuitive than `ClientCookie` and `ServerCookie`, but I'm happy to switch to these names instead.

Philip-NLnetLabs replied:
Clients send two types of cookies, the small one and the big. And receive only the big one. Servers receive small and big and send only big.

Another issue might be that if a client receives a server cookie, we cannot assume anything about what the server put there. The server might just send some random bytes.

Philip-NLnetLabs commented on src/new_rdata/basic.rs:
```diff
@@ -0,0 +1,531 @@
+#[repr(transparent)]
+pub struct Txt {
+    /// The text strings, as concatenated [`CharStr`]s.
+    content: [u8],
```
This seems wrong. A txt record is a collection of strings. Boundaries need to be preserved.

Philip-NLnetLabs replied:
Ah. Maybe an extra comment with what you just wrote.

ximon18 commented on src/new_edns/cookie.rs:
```diff
@@ -0,0 +1,244 @@
+    version: u8,
+
+    /// Reserved bytes in the cookie format.
+    reversed: [u8; 3],
```
That should be `reserved`, not `reversed`.

bal-e committed:
[new_base/wire/parse] Fix documentation typo

See: <https://github.com/NLnetLabs/domain/pull/474#discussion_r1925023434>

bal-e committed:
[new_edns/cookie] Rename 'CookieRequest' to 'ClientCookie'

Also fixes typo in field name 'reversed' ('reserved') of 'Cookie'.

See: <https://github.com/NLnetLabs/domain/pull/474#discussion_r1925050929>
See: <https://github.com/NLnetLabs/domain/pull/474#discussion_r1925115123>

bal-e committed:
[new_base/parse] Refactor '{Parse,Split}FromMessage'

- The traits have been renamed to '{Parse,Split}MessageBytes' for
  consistency with '{Parse,Split}Bytes'.

- The traits now consume 'message.contents' instead of the whole
  'message', allowing them to be used in more contexts (e.g. the
  upcoming overhauled builder types).

bal-e committed:
[new_base/build] Overhaul message building

- 'BuilderContext' now tracks the last question / record in a message,
  allowing its building to be recovered in the future (WIP).

- 'MessageBuilder' inlines the 'Builder' and just stores a mutable
  reference to a 'Message' for simplicity.

- 'Builder' no longer gives access to the message header, and uses
  '&UnsafeCell<[u8]>' to represent the message contents.

bal-e committed:
[new_base/build/message] Add resumption methods

bal-e committed:
[new_base/build] Test 'MessageBuilder' and fix some bugs

bal-e committed:
[new_base/build] Track section counts while building

bal-e committed:
[new_base/wire/parse] Fix miscount of doc test

bal-e committed:
Merge branch 'main' into new-base

bal-e committed:
[new_base/wire/parse] Support parsing into arrays

bal-e committed:
[new_base/build] Accept clippy suggestions

bal-e committed:
[new_rdata] Use 'core::net::Ipv4Addr'

bal-e committed:
[new_rdata] Document format in 'Txt'

See: <https://github.com/NLnetLabs/domain/pull/474#discussion_r1925085085>

bal-e committed:
[new_base/name] Add 'LabelBuf' and support building on 'Label'

bal-e committed:
[new_base/name] Add mutable access from 'RevNameBuf'

bal-e committed:
[new_base] Simply parse impls using 'split_*()'

I have taken care to preserve explicit parse impls for wrapper types
like 'SizePrefixed', and a few cases where performance could matter.

bal-e committed:
[new_base/name] Define the standard 'Name' type

bal-e committed:
[new_base/name] Add deconstruction methods for 'UnparsedName'

bal-e committed:
Remove accidental trailing whitespace

bal-e committed:
[new_base] Import various fixes from 'new-net-server'

bal-e committed:
[new_edns] Remove debug 'println'

bal-e committed:
[new_base] Add some basic unit tests

bal-e committed:
[new_rdata/basic] Refactor into submodules

bal-e committed:
[new_rdata] Impl equality for 'RecordData'

bal-e committed:
[new_rdata] Add 'map_name(s)_by_ref()'

bal-e committed:
Merge branch 'main' into new-base

Fixes up the existing codebase against new Clippy lints.

bal-e committed:
[new_base/wire] Make 'SizePrefixed' generic over size type

Some DNSSEC types use single-byte prefixed data, for which I'd still
like to make use of 'SizePrefixed' to avoid boilerplate.

bal-e committed:
[new_rdata] Implement stubs of DNSSEC record types

bal-e committed:
[src/new_base/size_prefixed] Fix clippy lint

bal-e committed:
[new_rdata/dnssec] Export helper types

bal-e committed:
[new_rdata] Avoid '{BuildInto,ParseFrom}Message' where possible

bal-e committed:
[new_rdata] Incorporate DNSSEC record types

bal-e committed:
[new_base/name/absolute] Add 'NameBuf'

This will be used for zonefile parsing.

bal-e committed:
[new_base] Add 'bumpalo' and impl 'clone_to_bump()'

'bumpalo' is useful for allocating unsized data types like record data.
It is currently used in the 'new-zonefile' branch as a buffer space for
records to be parsed into.  'clone_to_bump()' is necessary in order to
deep-copy record data into a new bump allocator.

bal-e committed:
[utils] Add 'UnsizedClone' and 'CloneFrom'

These are important for improving DST ergonomics; they make it much
easier to copy around unsized types, which 'domain' now has many of.

bal-e committed:
Define and use a derive macro for 'UnsizedClone'

bal-e committed:
[utils] Define 'clone_to_bump()' for integrating with 'bumpalo'

bal-e committed:
Replace most 'clone_to_bump()'s via 'UnsizedClone'

bal-e committed:
Fix broken doc links

bal-e committed:
[new_base/name/absolute] Impl 'Ord' for 'Name'

bal-e committed:
[new_base/message] Impl 'UnsizedClone' for 'Message'

bal-e committed:
[new_base/name] Introduce 'CanonicalName'

'CanonicalName' introduces important operations for DNSSEC.  It is
only implemented by 'Name' (not 'RevName') because 'RevName' cannot
implement it efficiently (users should prefer 'RecordData<&Name>').

bal-e committed:
Implement DNSSEC canonical ordering and building

bal-e committed:
[new_base/record] Fix broken doc link

bal-e committed:
[new_base] Add module 'compat'

bal-e committed:
[new_rdata/edns] Add 'Opt::EMPTY'

bal-e committed:
[new_base/build] Make 'MessageBuilder::finish()' return a mutable ref

bal-e committed:
[new_base/build] Add 'must_use' in a few vital places

This helps prevent users from accidentally forgetting to commit the
components of the message they're building.

bal-e committed:
[new_base/compat] Remove hard-to-port re-exports

bal-e committed:
[new_edns] Add 'EdnsRecord::clone_to_bump()'

bal-e committed:
[new_base/message] Make 'HeaderFlags' setters modify in place

bal-e committed:
[new_base/build] Fix broken doc test

bal-e committed:
[new_base/name] impl 'FromStr' for '*Buf' types

bal-e committed:
[new_base/wire] Clarify bits of doc test

bal-e committed:
[new_base/wire] Polish internal code

- Added some documentation and reasoning for the primitive types (U16
  etc.) provided in this module.

- Added documentation about the bounds on 'S' for 'SizePrefixed'.

- Improved the safety message for 'SizePrefixed'.

bal-e committed:
[new_base] Rephrase 'byte string' -> 'byte sequence'

- Along the way, the documentation for the parsing traits has been
  improved significantly.

bal-e committed:
[new_base/message] Rename 'slice_to()' to 'truncate()'

bal-e committed:
[new_base/message] Rename methods in 'HeaderFlags'

bal-e committed:
[utils] Simplify 'UnsizedClone' usage

bal-e committed:
[new_base/message] Fix wording of 'additional' in 'SectionCounts'

bal-e committed:
[new_base/serial] Use a custom method instead of 'Add'

Also updates it to disallow negative additions, as per the RFC.

bal-e committed:
[new_base/name/label] Allow wildcard labels in 'FromStr'

bal-e committed:
Merge branch 'main' into new-base

Account for conflicts in 'Cargo.lock' and 'lib.rs'.

ximon18 commented:
Unless I'm missing something the only way to make a `RevNameBuf` from a `RevName` is to do `let new_buf = RevNameBuf::copy_from(rev_name);` which feels weird. I tried doing `let new_buf = rev_name.clone()` but that didn't have the intended effect.

bal-e committed:
[new_base] Add more build/parse impls for 'std' wrappers

bal-e committed:
[new_rdata] Store 'A' and 'Aaaa' inline in 'RecordData'

bal-e committed:
[new_rdata] Expose 'dnssec::TypeBitmaps' publicly

bal-e committed:
[new_rdata/ipv6] Use 'core::net' instead of 'std::net'

bal-e committed:
[new_base/question] Adjust docs for 'QType' constants

bal-e committed:
[new_base] Add 'missing_doc' lints

bal-e committed:
[new_base/message] Thoroughly document 'HeaderFlags'

bal-e committed:
[new_base/name/unparsed] Allow 'UnparsedName' to be parsed from byte sequences

bal-e committed:
[new_base/record] Add 'Record::map_name()'

bal-e committed:
[new_base/record] Remove 'UnparsedRecord'

It's left over from experiments with high-level message parsing.

bal-e committed:
[new_base/record] Split 'ParseRecordDataBytes' from 'ParseRecordData'

This makes it possible to use 'RecordData::parse_record_data_bytes()'
with name types that don't support decompression.

bal-e committed:
[new_rdata] Add 'BoxedRecordData'

'BoxedRecordData' fills a missing bit of functionality, for long-term
storage of record data.  'RecordData' itself wasn't suitable for this
because of its lifetime parameter.

bal-e commented:
> Unless I'm missing something the only way to make a `RevNameBuf` from a `RevName` is to do `let new_buf = RevNameBuf::copy_from(rev_name);` which feels weird. I tried doing `let new_buf = rev_name.clone()` but that didn't have the intended effect. This is the pattern I would use with `&str` and `Path`. I see that the equivalent works for `Path` and `PathBuf` because `Clone` for `Path` is implemented in terms of `Path::to_path_buf()`. Something like that for `RevName` would be nice.

@ximon18: There are several possible container types for `RevName`: `RevNameBuf` and `Box<RevName>` are likely the most common.  `RevNameBuf` probably shouldn't be used too often (and even then, it should only be used as a local variable, not put on the heap), but that's not clear from the public API right now.  In any case, `RevName` implements `UnsizedClone`, which provides an `unsized_clone_into()` method that does what you're looking for, any supported container type.

bal-e committed:
[src/utils] impl 'CloneFrom' for 'Rc' and 'Arc'

Unlike 'Box', these types can't be constructed from a 'Layout'.  We
have to convince them to allocate in the right layout, which is hard
because we then need a compile-time known alignment.  I've added an
alignment type to 'UnsizedClone' for this.

bal-e committed:
[new_base/question] Remove 'UnparsedQuestion'

bal-e committed:
[new_rdata/basic/a] Document thoroughly

bal-e committed:
[utils/mod.rs] Import derive macro for 'UnsizedClone'

bal-e committed:
[new_rdata/basic/mx] Freshen up and document thoroughly

bal-e committed:
[new_rdata/basic/ns] Freshen up and document thoroughly

bal-e committed:
[new_rdata/basic/ptr] Freshen up and document thoroughly

bal-e committed:
[new_rdata/basic/ns] Adjust docs to specify authority

bal-e committed:
[new_rdata/basic/soa] Freshen up and document thoroughly

bal-e committed:
[utils::UnsizedClone] Fix doc test

bal-e committed:
[utils] Refactor 'UnsizedClone' -> 'dst::UnsizedCopy'

bal-e committed:
Refactor '{Parse,Split}BytesByRef' to '{Parse,Split}BytesZC'

- '*_by_ref()' methods have been renamed to '_zc()'.
- '*_by_mut()' methods are no longer offered; these will be offered
  through a separate trait, that will also support 'Box'.
- A new bound on 'UnsizedCopy' has been added, which provides
  'ptr_with_address()' (now 'ptr_with_addr()').

bal-e committed:
[new_base/parse/wire] Add 'ParseBytesInPlace'

bal-e committed:
[new_rdata/basic/txt] Clean up and document thoroughly

- Also introduces 'CharStrBuf' to simplify the building of 'Txt's.

bal-e committed:
[new_rdata/basic/cname] Clean up and document thoroughly

bal-e committed:
[new_rdata/basic/txt] Remove 'parse_boxed_bytes()'

This functionality is already provided by
'ParseBytesZC::parse_bytes_in()', with exactly the same signature.

bal-e committed:
[new_rdata] Remove 'Wks'

The existing 'base' module doesn't support it either.

bal-e committed:
[new_rdata/basic/hinfo] Clean up and document thoroughly

bal-e committed:
[new_base] Minor bugfixes (mainly documentation)

bal-e committed:
[new_base] Update doc tests to match new APIs

bal-e committed:
[new_rdata/ipv6] Clean up and document thoroughly

bal-e committed:
[new_base/wire/size_prefixed] Fix incorrect formatting

bal-e committed:
[new_base/wire/build] Add 'BuildBytes::built_bytes_size()'

This will make it much easier to allocate buffers for building into.
It is still preferable to over-allocate by a small amount rather than
to get the exact size.

Note that using 'built_bytes_size()' to reserve space in a 'Vec' is
almost certainly more efficient than incrementally appending to it, as
heap allocations are going to be more expensive than traversing any
object to calculate its serialization size.

bal-e committed:
[new_rdata/edns] Clean up and document thoroughly

- 'Opt' will now validate the option structure during parsing, acting
  like '[UnparsedEdnsOption]'.

- Introduced 'UnparsedEdnsOption', similar to 'UnparsedRecordData'.
  'Opt::options()' will return 'UnparsedEdnsOption' for every option
  it could not parse properly.

- Renamed 'UnknownOption' to 'UnknownOptionData' to distinguish how it
  works from 'UnparsedEdnsOption'.

bal-e committed:
Tweak some documentation

bal-e committed:
[new_*] Prefer 'alloc' over 'std'

bal-e committed:
[new_edns] Add conversion between 'Record' and 'EdnsRecord'

bal-e committed:
[new_base/parse] Add a high-level example

bal-e committed:
[new_base] Add 'MessageItem' and a mid-level parsing API

'Message's are semantically equivalent to sequences of 'MessageItem's,
and this is the basis for the mid-level API.  Next up: building.

bal-e committed:
[new_base] Expand top-level documentation

bal-e committed:
[new_base] Overhaul the building API

The mid-level building API has been completely rewritten.  It is much
simpler now, but no longer supports in-place modification or resumption
for the last message item.

'BuildIntoMessage' was renamed to 'BuildInMessage', and its signature
was completely changed to remove 'Builder'.  It is now quite similar to
'SplitMessageBytes'.  A slightly more involved truncation strategy is
used, where the name compressor is left to reset by the caller.

bal-e committed:
[new_base/build] Update doc test

bal-e committed:
[new_base] Implement name compresion

- Most file changes were to correct the parameter name for the name
  compressor in 'impl BuildInMessage::build_in_message()'.

- While the compressor seems to work correctly in basic cases, more
  thorough testing is necessary.

- It would be valuable to benchmark name compression between new-base
  and the existing base, e.g. in the construction of AXFR messages.

bal-e committed:
[new_base/name/compressor] Use case-insensitive hashing

bal-e committed:
Move 'new_*' to a dedicated 'new' module

This also adds an 'unstable-new' feature flag to gate these modules.

bal-e committed:
[new] Account for new Clippy lints

bal-e commented:
Final round of review before we hit merge tomorrow.

bal-e committed:
Merge branch 'main' into new-base

Updating to handle new Clippy lints in unrelated parts of the code.
